pkg/expr/mathexp/type_series.go

```diff
@@ -9,79 +9,113 @@ import (
 	"github.com/grafana/grafana/pkg/expr/mathexp/parse"
 )
 
-// Series has time.Time and ...? *float64 fields.
+// seriesTypeTimeIdx is the data frame field index for the Series type's Time column.
+const seriesTypeTimeIdx = 0
+
+// seriesTypeValIdx is the data frame field index for the Series type's Value column.
+const seriesTypeValIdx = 1
+
+// Series has a time.Time and a *float64 fields.
 type Series struct {
-	Frame           *data.Frame
-	TimeIsNullable  bool
-	TimeIdx         int
-	ValueIsNullable bool
-	ValueIdx        int
+	Frame *data.Frame
 	// TODO:
-	// - Multiple Value Fields
 	// - Value can be different number types
 }
 
 // SeriesFromFrame validates that the dataframe can be considered a Series type
-// and populate meta information on Series about the frame.
+// and mutates the frame to be in the format that additional SSE operations expect.
 func SeriesFromFrame(frame *data.Frame) (s Series, err error) {
 	if len(frame.Fields) != 2 {
 		return s, fmt.Errorf("frame must have exactly two fields to be a series, has %v", len(frame.Fields))
 	}
 
-	foundTime := false
-	foundValue := false
+	valueIdx := -1
+	timeIdx := -1
+
+	timeNullable := false
+	valueNullable := false
+
+FIELDS:
 	for i, field := range frame.Fields {
 		switch field.Type() {
 		case data.FieldTypeTime:
-			s.TimeIdx = i
-			foundTime = true
+			timeIdx = i
 		case data.FieldTypeNullableTime:
-			s.TimeIsNullable = true
-			foundTime = true
-			s.TimeIdx = i
+			timeNullable = true
+			timeIdx = i
 		case data.FieldTypeFloat64:
-			foundValue = true
-			s.ValueIdx = i
+			valueIdx = i
 		case data.FieldTypeNullableFloat64:
-			s.ValueIsNullable = true
-			foundValue = true
-			s.ValueIdx = i
+			valueNullable = true
+			valueIdx = i
 		default:
-			// Handle default case
+			if valueIdx != -1 && timeIdx != -1 {
+				break FIELDS
+			}
 		}
 	}
-	if !foundTime {
+
+	if timeIdx == -1 {
 		return s, fmt.Errorf("no time column found in frame %v", frame.Name)
 	}
-	if !foundValue {
+	if valueIdx == -1 {
 		return s, fmt.Errorf("no float64 value column found in frame %v", frame.Name)
 	}
+
+	if timeIdx != 0 { // make time field the first field
+		frame.Fields[seriesTypeTimeIdx], frame.Fields[timeIdx] = frame.Fields[timeIdx], frame.Fields[seriesTypeTimeIdx]
+	}
+
+	if valueIdx != 1 { // make value field the second field
+		frame.Fields[seriesTypeValIdx], frame.Fields[valueIdx] = frame.Fields[valueIdx], frame.Fields[seriesTypeValIdx]
+	}
```

Comment:
I think that this check will reverse the effect of the above block (timeIdx != 0).
Assuming that the value field is the first the 1st block will change the order of the fields but this one will change them again.